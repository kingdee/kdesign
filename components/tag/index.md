---
category: Components
type: 数据展示
order: 4
title: Tag
subtitle: 标签
---

标签是一种用于标记和选择的标识元素
## 使用场景
- 需要突出标记和选择时
- 视觉样式要与按钮严格区分

## API

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| clickable | 是否可以点击 | boolean | false | - | 1.0.0 |
| closable | 是否可关闭, 只有当 `type` 为 `edit` 时才生效 | boolean | false | - | 1.0.0 |
| closeIcon | 关闭按钮 | ReactNode | - | `<Icon type="close">` | 1.0.0 |
| color | 标签色.自定义颜色值，比如 `#f00` ,只有 `type` 为 `attribute` 时才生效 | string | `process` | `process` `success` `warning` `error` `end` `expired` | 1.0.0 |
| disabled | 是否禁用, 只有当 `type` 为 `edit` 时才生效 | boolean | `false` | - | 1.0.0 |
| icon | 设置图标 | ReactNode | - | `<Icon type="tips">` | 1.0.0 |
| size | 标签大小 | string | `middle` | `small` `middle` `large` | 1.0.0 |
| type | 标签类型 | string | `status` | `status` `attribute` `text` `edit` | 1.0.0 |
| onClick | 点击时的回调,只有当 `clickable` 设置为 `true` 时才能触发 | function(e:Event) | - | - | 1.0.0 |
| onClose | 关闭时的回调 | function(e:Event) | - | - | 1.0.0 |

## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-tag-color-process | --kd-g-color-ongoing | #276ff5 |
|  | --kd-c-tag-color-success | --kd-g-color-success | #1ba854 |
|  | --kd-c-tag-color-warning | --kd-g-color-warning | #ff991c |
|  | --kd-c-tag-color-error | --kd-g-color-error | #fb2323 |
|  | --kd-c-tag-color-end | --kd-g-color-end | #666 |
|  | --kd-c-tag-color-expired | --kd-g-color-disabled | #b2b2b2 |
|  | --kd-c-tag-color-disabled | --kd-g-color-disabled | #b2b2b2 |
|  | --kd-c-tag-attribute-color-text | --kd-g-color-white | #fff |
|  | --kd-c-tag-edit-color-border | --kd-g-color-text-placeholder | #ccc |
|  | --kd-c-tag-edit-color-text | --kd-g-color-text-primary | #212121 |
|  | --kd-c-tag-edit-color-border-hover | --kd-g-color-theme | #5582f3 |
|  | --kd-c-tag-edit-cloesWrapper-color-background-hover | --kd-g-color-theme | #5582f3 |
|  | --kd-c-tag-edit-cloesWrapper-color-text-hover | --kd-g-color-theme | #5582f3 |
|  | --kd-c-tag-edit-color-background-disabled | --kd-g-color-background-contain-disabled | #f5f5f5 |
|  | --kd-c-tag-edit-color-background-disabled | --kd-g-color-border-strong | #d9d9d9 |
|  | --kd-c-tag-edit-color-text-disabled | --kd-g-color-border-strong | #d9d9d9 |
| font | --kd-c-tag-font-size-small | --kd-g-font-size-small | 12px |
|  | --kd-c-tag-font-size-middle | --kd-g-font-size-small | 12px |
|  | --kd-c-tag-font-size-large | --kd-g-font-size-middle | 14px |
| motion | --kd-c-tag-motion-duration | --kd-g-duration | 0.3s |
| radius | --kd-c-tag-radius-border | - | 10px |
| sizing | --kd-c-tag-sizing-height-small | - | 20px |
|  | --kd-c-tag-sizing-height-middle | - | 20px |
|  | --kd-c-tag-sizing-height-large | - | 24px |
|  | --kd-c-tag-sizing-max-width | - | - |
|  | --kd-c-tag-sizing-min-width | - | - |
| spacing | --kd-c-tag-spacing-padding-horizontal-small | - | 6px |
|  | --kd-c-tag-spacing-padding-horizontal-middle | - | 7px |
|  | --kd-c-tag-spacing-padding-horizontal-large | - | 8px |
