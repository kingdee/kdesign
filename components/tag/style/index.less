
@import '../../style/themes/index';
@import './mixin.less';
@tag-prefix-cls: ~'@{kd-prefix}-tag';

@types: process, success, warning, error, end, expired;
@colors: @tag-process-color, @tag-success-color, @tag-warning-color, @tag-error-color, @tag-end-color, @tag-expired-color;

.@{tag-prefix-cls} {
  &-ellipsis {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  &-size-small {
    .tag-size(@tag-small-font-size, @tag-small-height, @tag-small-padding-horizontal);
  }

  &-size-middle {
    .tag-size(@tag-middle-font-size, @tag-middle-height, @tag-middle-padding-horizontal);
  }

  &-size-large {
    .tag-size(@tag-large-font-size, @tag-large-height, @tag-large-padding-horizontal);
  }

  transition: all @tag-g-motion-duration;
  // 状态标签
  &-shape-status {
    .tag-status(@tag-process-color)
  }
  each(@types, {
      &-shape-status.@{tag-prefix-cls}-@{value} {
        .tag-status(extract(@colors, @index))
      }
  })

  // 属性标签
  &-shape-attribute {
    .tag-attribute(@tag-process-color)
  }
  each(@types, {
    &-shape-attribute.@{tag-prefix-cls}-@{value} {
      .tag-attribute(extract(@colors, @index))
    }
  })
  // 文本标签
  &-shape-text {
    .tag-text-color(@tag-process-color);
  }
  each(@types, {
    &-shape-text.@{tag-prefix-cls}-@{value} {
      .tag-text-color(extract(@colors, @index))
    }
    &-shape-text.@{tag-prefix-cls}-@{value}.@{tag-prefix-cls}-size-middle {
      .tag-text-color(extract(@colors, @index))
    }
    &-shape-text.@{tag-prefix-cls}-@{value}.@{tag-prefix-cls}-size-large {
      .tag-text-color(extract(@colors, @index))
    }
  })

  // 可关闭的标签
  &-shape-edit {
    border: 1px solid @tag-edit-g-color-border;
    border-radius: 2px;
    line-height: @tag-middle-height;
    padding-left: 6px;
    padding-right: 6px;
    color: @tag-edit-g-text-color;
    &:not(.@{tag-prefix-cls}-closable-disabled):hover {
      border-color: @tag-edit-g-color-border-hover;
      .@{tag-prefix-cls}-closeWrapper {
        // background-color: @tag-edit-cloesWrapper-g-color-background-hover;
        color: @tag-edit-cloesWrapper-g-text-color-hover;
      }
    }

    &.@{tag-prefix-cls}-closable {
      padding-right: 8px;
    }
  }

  // 可点击的标签
  &-clickable {
    cursor: pointer;
  }

  // 自定义标签颜色
  &-has-color {
    color: @tag-attribute-g-text-color;
  }

  // 图标设置外边距
  &-text {
    margin-left: 6px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  &-closable-disabled {
    background-color: @tag-edit-g-color-background-disabled;
    border-color: @tag-edit-g-color-border-disabled;
    color: @tag-edit-g-text-color-disabled;
    cursor: not-allowed;
  }
  &-closable.@{tag-prefix-cls}-size-middle {
    line-height: @tag-middle-height;
  }
  &-closable.@{tag-prefix-cls}-size-large {
    line-height: @tag-large-height;
    padding-left: @tag-large-padding-horizontal;

    .@{tag-prefix-cls}-closeWrapper {
      .@{tag-prefix-cls}-close-icon {
        font-size: 14px;
      }
    }
  }
  &-closeWrapper {
    margin-left: 4px;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    transition: color @tag-g-motion-duration, background-color @tag-g-motion-duration;
  }
  &-close-icon {
    cursor: pointer;
    font-size: 12px;
  }
}
