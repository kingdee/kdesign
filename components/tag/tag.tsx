import React, { FunctionComponentElement, useContext, useRef } from 'react'
import { Icon } from '../index'
import classNames from 'classnames'
import ConfigContext from '../config-provider/ConfigContext'
import { tuple } from '../_utils/type'
import { getCompProps } from '../_utils'
import devWarning from '../_utils/devwarning'

export const TagTypes = tuple('status', 'attribute', 'text', 'edit')
export type TagType = typeof TagTypes[number]
export const TagColors = tuple('process', 'success', 'warning', 'error', 'end', 'expired')
export type TagColor = typeof TagColors[number]
export const TagSizes = tuple('small', 'middle', 'large')
export type TagSize = typeof TagSizes[number]

export interface ITagProps {
  className?: string // 样式名
  clickable?: boolean // 是否可点击
  closable?: boolean // 是否可关闭
  closeIcon?: React.ReactNode // 关闭按钮
  color?: TagColor | string // 标签颜色 process success warning error end expired 也可以传入色值
  disabled?: boolean
  icon?: React.ReactNode // 设置图标
  size?: TagSize
  style?: React.CSSProperties
  type?: TagType // 标签类型
  onClick?: React.MouseEventHandler<HTMLElement> // 点击时的回调
  onClose?: React.MouseEventHandler<HTMLElement> // 关闭时的回调
  children?: React.ReactNode // 子元素
  title?: string
}

// remove this line and code Tag component here
const InteranalTag = (props: ITagProps, ref: unknown): FunctionComponentElement<any> => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = useContext(ConfigContext)
  const tagProps = getCompProps('Tag', userDefaultProps, props)
  const {
    className,
    clickable,
    closable,
    closeIcon,
    color,
    disabled,
    icon,
    size,
    type,
    onClick,
    onClose,
    style,
    prefixCls: customPrefixcls,
    children,
    ...others
  } = tagProps

  devWarning(TagTypes.indexOf(type) === -1, 'tag', `cannot found tag type '${type}'`)
  devWarning(TagSizes.indexOf(size) === -1, 'tag', `cannot found tag size '${size}'`)

  const tagPrefixCls = getPrefixCls!(prefixCls, 'tag', customPrefixcls)

  const handleClick = (e: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
    if (!clickable) return
    onClick && onClick(e)
  }

  const handleClose = (e: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
    onClose && onClose(e)
    if (e.defaultPrevented) {
      return
    }
    // 删除标签
    const tagWrapper = e.currentTarget.parentNode
    tagWrapper?.parentNode?.removeChild(tagWrapper)
  }

  // ref
  const thisTagRef = useRef<HTMLElement>()
  const tagRef = (ref as any) || thisTagRef

  // 对children进行进一步处理 当标签包含图标时 文本内容需要用标签包裹设置外边距
  const handleChild = function (child: React.ReactChild) {
    if (!child) return null
    if (typeof child === 'string' && icon) {
      return <span className={`${tagPrefixCls}-text`}>{child}</span>
    }
    return <span className={`${tagPrefixCls}-ellipsis`}>{child}</span>
  }

  // 预设的颜色值
  const isPresetColor = ['process', 'success', 'warning', 'error', 'end', 'expired'].indexOf(color) > -1

  const tagClasses = classNames(tagPrefixCls, className, {
    [`${tagPrefixCls}-shape-${type}`]: type,
    [`${tagPrefixCls}-${color}`]: isPresetColor,
    [`${tagPrefixCls}-has-color`]: !isPresetColor && color && type === 'attribute',
    [`${tagPrefixCls}-clickable`]: clickable,
    [`${tagPrefixCls}-closable`]: closable,
    [`${tagPrefixCls}-size-${size}`]: size,
    [`${tagPrefixCls}-closable-disabled`]: type === 'edit' && disabled,
  })

  // 设置自定义颜色值的样式
  const tagColorStyle =
    !isPresetColor && color && type === 'attribute'
      ? {
          backgroundColor: color,
        }
      : {}

  const tagStyle = {
    ...tagColorStyle,
    ...style,
  }

  // 图标
  const iconNode = icon || null

  const tagChild = handleChild(children)

  // 关闭按钮
  const closeIconNode = closeIcon || <Icon type="close" className={`${tagPrefixCls}-close-icon`} />
  const closeNode =
    closable && type === 'edit' ? (
      disabled ? (
        <span className={`${tagPrefixCls}-closeWrapper`}>{closeIconNode}</span>
      ) : (
        <span className={`${tagPrefixCls}-closeWrapper`} onClick={handleClose}>
          {closeIconNode}
        </span>
      )
    ) : null

  return (
    <span ref={tagRef} onClick={handleClick} className={tagClasses} style={tagStyle} {...others}>
      {iconNode}
      {tagChild}
      {closeNode}
    </span>
  )
}

const Tag = React.forwardRef<unknown, ITagProps>(InteranalTag)
Tag.displayName = 'Tag'
export default Tag
