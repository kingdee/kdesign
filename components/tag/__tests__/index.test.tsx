import React from 'react'
import { render, mount } from 'enzyme'
import Tag, { TagType, TagSize } from '../index'
import { TagTypes, TagColors, TagSizes } from '../tag'
import Icon from '../../icon'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'

describe('Tag', () => {
  // 1. mount test
  mountTest(Tag)
  TagTypes.forEach((type) => {
    mountTest(() => <Tag type={type} />)
  })

  // 2. render test
  it('renders correctly', () => {
    expect(render(<Tag>Text</Tag>)).toMatchSnapshot()

    TagTypes.forEach((type) => {
      expect(render(<Tag type={type}>Text</Tag>)).toMatchSnapshot()
    })

    TagColors.forEach((color) => {
      expect(render(<Tag color={color}>Text</Tag>)).toMatchSnapshot()
    })

    TagSizes.forEach((size) => {
      expect(render(<Tag size={size}>Text</Tag>)).toMatchSnapshot()
    })
  })

  // 3. warns in component
  it('warns if type is wrong', () => {
    const mockWarn = jest.fn()
    jest.spyOn(console, 'warn').mockImplementation(mockWarn)
    const type = 'who am I' as any as TagType
    render(<Tag type={type} />)
    expect(mockWarn).toHaveBeenCalledTimes(1)
    expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-tag: cannot found tag type 'who am I'")

    const size = 'who am I' as any as TagSize
    render(<Tag size={size} />)
    expect(mockWarn).toHaveBeenCalledTimes(2)
    expect(mockWarn.mock.calls[1][0]).toMatch("Warning: [kdesign]-tag: cannot found tag size 'who am I'")
  })

  // 4. render null or undefined without errors
  it('render null or undefined tag without errors', () => {
    expect(
      mount(
        <Tag>
          {null}
          {undefined}
        </Tag>,
      ).find('.kd-tag-ellipsis').length,
    ).toBe(1)
  })

  // 5. displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Tag>Text</Tag>)
    expect((wrapper.type() as any).displayName).toBe('Tag')
  })

  // 6. class state
  it('should class use right', () => {
    const DefaultTag = mount(<Tag>Text</Tag>)
    expect(DefaultTag.find('.kd-tag')).toHaveClassName('.kd-tag-shape-status')

    TagTypes.forEach((type) => {
      const TestTag = mount(<Tag type={type}>Text</Tag>)
      expect(TestTag.find(`.kd-tag`)).toHaveClassName(`.kd-tag-shape-${type}`)
    })

    TagColors.forEach((color) => {
      const TestTag = mount(<Tag color={color}>Text</Tag>)
      expect(TestTag.find(`.kd-tag`)).toHaveClassName(`.kd-tag-${color}`)
    })

    TagSizes.forEach((size) => {
      const TestTag = mount(<Tag size={size}>Text</Tag>)
      expect(TestTag.find(`.kd-tag`)).toHaveClassName(`.kd-tag-size-${size}`)
    })

    const CustomColorTag = mount(
      <Tag type="attribute" color="#ff0">
        Text
      </Tag>,
    )
    expect(CustomColorTag.find(`.kd-tag`)).toHaveClassName(`.kd-tag-has-color`)

    const IconTag = mount(<Tag icon={<Icon type="add-solid" />}>Text</Tag>)
    expect(IconTag.find(`.kd-tag`)).toContainReact(<Icon type="add-solid"></Icon>)

    const ClickableTag = mount(<Tag clickable>Text</Tag>)
    expect(ClickableTag.find(`.kd-tag`)).toHaveClassName(`.kd-tag-clickable`)

    const ClosableTag = mount(<Tag closable>Text1</Tag>)
    expect(ClosableTag.find(`.kd-tag`)).toHaveClassName(`.kd-tag-closable`)

    const DisabledTag = mount(
      <Tag type="edit" closable disabled>
        Text1
      </Tag>,
    )
    expect(DisabledTag.find(`.kd-tag`)).toHaveClassName(`.kd-tag-closable-disabled`)
  })

  // 7.component interaction(event)
  it('should clickable ', () => {
    const onClick = jest.fn()
    const wrapper = mount(
      <Tag clickable onClick={onClick}>
        text
      </Tag>,
    )
    wrapper.simulate('click')
    expect(onClick).toHaveBeenCalled()
  })

  it('should closable ', () => {
    const onClose = jest.fn()
    const wrapper = mount(
      <Tag type="edit" closable onClose={onClose}>
        text
      </Tag>,
    )
    wrapper.find(`.kd-tag-closeWrapper`).simulate('click')
    expect(onClose).toHaveBeenCalled()
  })

  it('should not closable ', () => {
    const onClose = jest.fn()
    const wrapper = mount(
      <Tag type="edit" closable onClose={onClose} disabled>
        text
      </Tag>,
    )
    wrapper.find(`.kd-tag-closeWrapper`).simulate('click')
    expect(onClose).toHaveBeenCalledTimes(0)
  })

  // 8.config provider
  it('should config use config provider', () => {
    const tagConfig = {
      compDefaultProps: {
        Tag: {
          type: 'attribute',
        },
      },
    }
    const wrapper = mount(
      <ConfigProvider value={tagConfig}>
        <Tag>Tag Text</Tag>
      </ConfigProvider>,
    )
    expect(wrapper.find('.kd-tag')).toHaveClassName('.kd-tag-shape-attribute')
  })

  // 9. ref test
  it('should get Tag element from ref', () => {
    const ref = React.createRef()
    mount(<Tag ref={ref} type="status"></Tag>)
    expect(ref.current instanceof HTMLSpanElement).toBe(true)
  })

  // 10. api test
  it('api test', () => {
    // closable disabled color
    TagTypes.forEach((type) => {
      const tagDemo = mount(<Tag type={type} closable color="#f00" disabled></Tag>)
      if (type === 'edit') {
        expect(tagDemo.find('.kd-tag .kd-tag-close-icon')).toExist()
        expect(tagDemo.find('.kd-tag.kd-tag-closable-disabled')).toExist()
      } else {
        expect(tagDemo.find('.kd-tag .kd-tag-close-icon')).not.toExist()
        expect(tagDemo.find('.kd-tag.kd-tag-closable-disabled')).not.toExist()
      }
      // color
      if (type === 'attribute') {
        expect(tagDemo.find('.kd-tag.kd-tag-has-color').props()?.style?.backgroundColor).toEqual('#f00')
      } else {
        expect(tagDemo.find('.kd-tag').props()?.style?.backgroundColor).not.toEqual('#f00')
      }
    })

    // closeIcon
    const tagDemo = mount(<Tag type="edit" closable closeIcon={<Icon type="add"></Icon>}></Tag>)
    expect(tagDemo.find('.kd-tag .kdicon-add')).toExist()
  })
})
