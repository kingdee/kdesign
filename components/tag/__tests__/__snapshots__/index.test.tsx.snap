// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders correctly 1`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 2`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 3`] = `
<span
  class="kd-tag kd-tag-shape-attribute kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 4`] = `
<span
  class="kd-tag kd-tag-shape-text kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 5`] = `
<span
  class="kd-tag kd-tag-shape-edit kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 6`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-process kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 7`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-success kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 8`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-warning kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 9`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-error kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 10`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-end kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 11`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-expired kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 12`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-size-small"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 13`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-size-middle"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;

exports[`Tag renders correctly 14`] = `
<span
  class="kd-tag kd-tag-shape-status kd-tag-size-large"
>
  <span
    class="kd-tag-ellipsis"
  >
    Text
  </span>
</span>
`;
